core/src/nucleus/ribosome/fn_call.rs

```diff
@@ -0,0 +1,1271 @@
+/// fn_call is the module that implements calling zome functions
+///
+use crate::{
+    action::{Action, ActionWrapper},
+    context::Context,
+    instance::{dispatch_action_with_observer, Observer},
+    nucleus::{actions::get_entry::get_entry_from_agent_chain, ribosome::{self, WasmCallData}, state::NucleusState},
+};
+use holochain_core_types::{
+    cas::content::Address,
+    dna::{
+        capabilities::{CallSignature, CapabilityCall, CapabilityType},
+        wasm::DnaWasm,
+    },
+    entry::{cap_entries::CapTokenGrant, Entry},
+    error::{HcResult, HolochainError},
+    json::JsonString,
+    signature::Signature,
+};
+use snowflake;
+use std::{
+    sync::{
+        mpsc::{sync_channel, SyncSender},
+        Arc,
+    },
+    thread,
+};
+
+/// Struct holding data for requesting the execution of a Zome function (ExecuteZomeFunction Action)
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct ZomeFnCall {
+    id: snowflake::ProcessUniqueId,
+    pub zome_name: String,
+    pub cap: CapabilityCall,
+    pub fn_name: String,
+
+    pub parameters: JsonString,
+}
+
+impl ZomeFnCall {
+    pub fn new<J: Into<JsonString>>(
+        zome: &str,
+        cap: CapabilityCall,
+        function: &str,
+        parameters: J,
+    ) -> Self {
+        ZomeFnCall {
+            // @TODO can we defer to the ActionWrapper id?
+            // @see https://github.com/holochain/holochain-rust/issues/198
+            id: snowflake::ProcessUniqueId::new(),
+            zome_name: zome.to_string(),
+            cap: cap,
+            fn_name: function.to_string(),
+            parameters: parameters.into(),
+        }
+    }
+
+    pub fn create<J: Into<JsonString>>(
+        context: Arc<Context>,
+        zome: &str,
+        token: Address,
+        caller: Address,
+        function: &str,
+        parameters: J,
+    ) -> Self {
+        let params = parameters.into();
+        ZomeFnCall::new(
+            zome,
+            make_cap_call(context, token, caller, function, params.clone()),
+            function,
+            params,
+        )
+    }
+
+    pub fn same_fn_as(&self, fn_call: &ZomeFnCall) -> bool {
+        self.zome_name == fn_call.zome_name
+            && self.cap == fn_call.cap
+            && self.fn_name == fn_call.fn_name
+    }
+
+    pub fn cap_token(&self) -> Address {
+        self.cap.cap_token.clone()
+    }
+}
+
+/// Reduce ExecuteZomeFunction Action
+/// Execute an exposed Zome function in a separate thread and send the result in
+/// a ReturnZomeFunctionResult Action on success or failure
+pub(crate) fn reduce_execute_zome_function(
```

Comment:
😱 
Is this whole file an accidental add?

Reply:
yep.